app/Http/Controllers/Admin/AuthorController.php

```diff
@@ -4,19 +4,27 @@
 
 use App\Http\Controllers\Controller;
 use App\Http\Requests\AuthorRequest;
+use App\Repositories\Author\AuthorRepositoryInterface;
 use App\Models\Author;
 use Illuminate\Http\Request;
 
 class AuthorController extends Controller
 {
+    protected $authorRepo, $bookRepo;
+
+    public function __construct(
+        AuthorRepositoryInterface $authorRepo
+    ){
+       $this->authorRepo = $authorRepo;
+    }
     /**
      * Display a listing of the resource.
      *
      * @return \Illuminate\Http\Response
      */
     public function index()
     {
-        $authors = Author::orderBy('id', 'DESC')->paginate(config('pagination.limit_page'));
+        $authors = $this->authorRepo->getAll();
```

Comment:
repository có ưu điểm, nhược điểm gì?

Reply:
Repo the em thấy nó có ưu điểm là:
1. Tối ưu hóa được code 
2. Khi mình thay đổi database thì cũng dễ dàng thay đổi hơn ạ 
3. Em đọc bài của chị thì thấy nó dễ dàng test hơn ạ 
Nhược điểm :
1. Cần chọn ra những hàm nào riêng của class để viết vào Repo riêng của class đó
VD: class ```author``` có 1 hàm riêng để lấy ra quan hệ với ```book``` ạ
2. Cần nhiều kiến thức hơn để viết repo vì nó không dễ dàng viết như 1 controller bình thường ạ

Reply:
> Khi mình thay đổi database thì cũng dễ dàng thay đổi hơn ạ

thay đổi DB chỉ là 1 ví dụ về phụ thuộc. mở rộng ra khi logic thay đổi hoặc phải dùng đến cả file logic mới
> Cần chọn ra những hàm nào riêng của class để viết vào Repo riêng của class đó

Cái này không phải là nhược điểm mà là 1 ưu điểm, tới ưu và reuse code
> Cần nhiều kiến thức hơn để viết repo vì nó không dễ dàng viết như 1 controller bình thường ạ

Repo không hề khó viết. chỉ là cấu trúc folder sẽ lớn hơn thôi.

Reply:
Khi lần đầu viết ```Repo``` thì em thấy là một khung mẫu cho mình viết code một cách nhanh chóng, thống nhất code, khi mình sử dụng các hàm của ```Repo``` :D